src/validation.ts

```diff
@@ -371,4 +524,138 @@ export class OpenAPIValidator {
     const validator = new Ajv(this.ajvOpts);
     this.responseValidators[operationId] = validator.compile(schema);
   }
+
+  /**
+   * Get response validator function for an operation by operationId
+   *
+   * @param {string} operationId
+   * @returns {Object.<Ajv.ValidateFunction>}}
+   * @memberof OpenAPIRequestValidator
+   */
+  public getStatusBasedResponseValidatorForOperation(operationId: string): StatusBasedResponseValidatorsFunctionMap {
+    return this.statusBasedResponseValidators[operationId];
+  }
+
+  /**
+   * Builds an ajv response validator function for an operation and registers it to responseHeadersValidators
+   *
+   * @param {Operation} operation
+   * @memberof OpenAPIRequestValidator
+   */
+  public buildStatusBasedResponseValidatorForOperation(operation: Operation): void {
+    if (!operation.responses) {
+      // operation has no responses, don't register a validator
+      return null;
+    }
+
+    const { operationId } = operation;
+    const responseValidators: StatusBasedResponseValidatorsFunctionMap = {};
+    const validator = new Ajv(this.ajvOpts);
+
+    _.mapKeys(operation.responses, (response: OpenAPIV3.ResponseObject, status: string) => {
+      if (response.content && response.content['application/json'] && response.content['application/json'].schema) {
+        const validateFn = response.content['application/json'].schema;
+        responseValidators[status] = validator.compile(validateFn);
+      }
+      return null;
+    });
+
+    this.statusBasedResponseValidators[operationId] = responseValidators;
+  }
+
+  /**
+   * Get response validator function for an operation by operationId
+   *
+   * @param {string} operationId
+   * @returns {Object.<Object.<Ajv.ValidateFunction>>}}
+   * @memberof OpenAPIRequestValidator
+   */
+  public getResponseHeadersValidatorForOperation(operationId: string): ResponseHeadersValidateFunctionMap {
+    return this.responseHeadersValidators[operationId];
+  }
+
+  /**
+   * Builds an ajv response validator function for an operation and registers it to responseHeadersValidators
+   *
+   * @param {Operation} operation
+   * @memberof OpenAPIRequestValidator
+   */
+  public buildResponseHeadersValidatorForOperation(operation: Operation): void {
+    if (!operation.responses) {
+      // operation has no responses, don't register a validator
+      return null;
+    }
+
+    const { operationId } = operation;
+    const headerValidators: ResponseHeadersValidateFunctionMap = {};
+    const validator = new Ajv({ ...this.ajvOpts, coerceTypes: true });
+
+    _.mapKeys(operation.responses, (response: OpenAPIV3.ResponseObject, status: string) => {
```

Comment:
I'm not lodash expert (far from it), but mapping is probably not what you're looking to do. A map operation takes array with length N and returns a new array with length N. In this case, this array is full of `null` values, as that's what you return. Rather wasteful. You really seem to want to iterate over the keys, not map them. Same is true for other mapKeys usage in this PR, but I won't point them out individually :)

Reply:
Exactly yes, the `mapKeys` is used only to produce a side effect in the lambda passed as second argument.

I would love to avoid to waste that memory allocation and put unnecessary pressure on the GC, especially with a lot of items in the array, the allocated (and wasted) array will get bigger and bigger. However I follow the philosophy in place.

@anttiviljami Would you allow me to rewrite such code with a mere `for` loop ? What about all other places in order to keep the code consistent ?

Reply:
Ah, this was already a pattern in the codebase? Would be nice to address that. To be honest, I think most lodash can easily be replaced by relatively modern JS. But yes, in the very least it would be good for performance/GC if these "misuses" (sorry) of `mapKeys` would be replaced with something more appropriate.

Reply:
Good point! I'd be happy if we could get rid of unnecessary lodash stuff like mapKeys. Just an old habit I've yet to get rid of. Didn't think it would make a big difference. :)

Super grateful if anyone could submit a PR to rewrite my mapKeys usage to more modern for...in loops?

Reply:
For this PR. I'm fine with keeping the mapKeys as is. Let's change that later when we get to it.

Reply:
Alright, thanks.

Reply:
I made a [JSPerf](https://jsperf.com/mapkeysvsmodernjs/1) just to check the impact.
Its 3 times faster to just loop with a `for..in`.

Reply:
I added a few variations for your enjoyment: https://jsperf.com/mapkeysvsmodernjs/3
Fun to see for..in is indeed the fastest of all (in Chrome, which is the JS engine we care about).